### ISSUE TITLE
Celery producer hight memory usage with redis broker

### ISSUE THREAD

**Issue by p-eli** [celery/celery#5391]
Hi, 
I use celery in distributed system and I have problem with high memory usage. Probably it's not bug but I only do something wrong. It's look like application store data in some cache. Because, if I produce two same task(group/chord) memory usage grown up only after the first one.

I have tornado server which have two endpoinds - one produce task set and save tasks ID to database. Second load tasks id from database and check if tasks are finished.  Task consumers are running separate.  Backed and Broker is redis.

Application is running in multiple instance, so set and read tasks result from one application instance it's not guaranteed --> we must store task id in database. 

Use version:
 - celery 4.2.1
 - kombu 4.4.0
 - amqp 2.4.2
 - redis 3.2.0
 - python 3.5.3

Producer celery setting:
```
        celery_app = Celery()
        celery_app.conf.broker_url = "sentinel://abc:26379/0;sentinel://def:26379/0"
        celery_app.conf.result_backend = "sentinel://abc:26379/1;sentinel://def:26379/1"
        celery_app.conf.task_create_missing_queues = True
        celery_app.conf.broker_transport_options = {'master_name': 'some_master', 'socket_timeout': 10, 'socket_keepalive': True, 'retry_on_timeout': True}
        celery_app.conf.result_backend_transport_options = {'master_name': 'some_master', 'socket_timeout': 10, 'socket_keepalive': True, 'retry_on_timeout': True}
        celery_app.conf.broker_password = 'broker_password'
        celery_app.conf.redis_password = 'redis_password'
```

Task set producer:
```
    task = chain(
                        group(
                            signature('tasks.some_task_1', args=[1, 1, 1], task_id=str(x)) for x in range(0, 50000)
                        ),
                        signature('tasks.some_task_1', args=[1], task_id="specific_id")
                )
     task.apply_async()
     # here job finish -- we never need work with AsyncResult object 
````

How to reproduce:
  1. Run task set producer
       - memory has increased by ~200MiB
  2. Run task set producer again
       - memory doesn't increased again
  3. Memory has been never free again
    

Main Problem:
 1. If I produce big set of task, memory has increased to memory limit and hold 
         - producer working without problem - free some memory if producer produce new set of task 
            - e.g. if memory limit is 200 memory check in time:
                     20/200 MiB  -- before first task set
                       ....
                     199/200 MiB  -- after first task set
                     199/200 MiB  -- hold
                     185/200 MiB  -- free some memory for second task set
                     199/200 MiB  -- produce second task set
                       ....
  --> producing task set work but application crashed if need some memory for do another stuff  
                      
Memory profile output:
```
{"alloc-entries": [{"id": "/root/.cache/pypoetry/virtualenvs/produce-app/lib/python3.5/site-packages/celery/backends/base.py:579", "count": 20000, "size": 1077780},
 {"id": "<frozen importlib._bootstrap_external>:484", "count": 6018, "size": 556282},
 {"id": "/root/.cache/pypoetry/virtualenvs/produce-app/lib/python3.5/site-packages/celery/backends/redis.py:136", "count": 1, "size": 524288},
 {"id": "/usr/lib/python3.5/weakref.py:168", "count": 2, "size": 393664},
 {"id": "/root/.cache/pypoetry/virtualenvs/produce-app/lib/python3.5/site-packages/redis/client.py:3101", "count": 1, "size": 393248},
 {"id": "/root/.cache/pypoetry/virtualenvs/produce-app/lib/python3.5/site-packages/kombu/transport/virtual/base.py:916", "count": 4, "size": 267048},
 {"id": "/root/.cache/pypoetry/virtualenvs/produce-app/lib/python3.5/site-packages/redis/client.py:3117", "count": 1, "size": 262144},
 {"id": "/root/.cache/pypoetry/virtualenvs/produce-app/lib/python3.5/site-packages/celery* Connection #0 to host localhost left intact
/canvas.py:171", "count": 1992, "size": 143424},
 {"id": "/root/.cache/pypoetry/virtualenvs/produce-app/lib/python3.5/site-packages/celery/utils/collections.py:435", "count": 2134, "size": 137728},
 {"id": "/usr/lib/python3.5/collections/__init__.py:431", "count": 811, "size": 59284},
 {"id": "remaining-items-grouped", "count": null, "size": 1089551}], "status-msg": null, "status": "OK", "memory-profiler": "tracemalloc", "report-grouping-type": "lineno", "total-alloc-bytes": 4904441}
```

My question:
- It's some option to free allocated memory or don't cached data?  Or some clean up mechanism? Or some mechanism for only pushing task? 

Sorry that I don't use your issue template. If my problem is really bug a can create new issue with template. 

Thank you

**Comment by SHxKM** [comment]
I'm also looking for a solution to this. Using `--max-tasks-per-child` and `--auto-scale` helps kill workers and free some memory periodically but it still hovers over 100% at one particular chord-task.

**Comment by auvipy** [comment]
closing this for now. feel free to profile your production system and related celery codes. and share with the community.

**Comment by dromer** [comment]
We have been experiencing extremely high memory usage from using redis as a broker, this does not seem to be resolved to me.

- redis: 5.0.9
- celery: 5.0.5

As you can see the memory usage just keeps going up steadily. We are investigating if there are some redis settings we can tweak to mitigate this.

![](https://i.imgur.com/2BRhuPD.png)

**Comment by dmelo** [comment]
@dromer , were you able to figure out how to solve this issue? I'm experiencing high memory usage on Redis too.

**Comment by dromer** [comment]
@dmelo no, we moved back to rabbitmq and all was well.

I no longer work on this project.
